In [0]:
import tensorflow
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from pandas import read_csv

In [87]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


1. Read the dataset

In [88]:
# load dataset

dataframe = read_csv("/content/drive/My Drive/Colab Notebooks/aiml/bank.csv")
dataframe.head()

RowNumber  CustomerId   Surname  ...  IsActiveMember EstimatedSalary Exited
0          1    15634602  Hargrave  ...               1       101348.88      1
1          2    15647311      Hill  ...               1       112542.58      0
2          3    15619304      Onio  ...               0       113931.57      1
3          4    15701354      Boni  ...               0        93826.63      0
4          5    15737888  Mitchell  ...               1        79084.10      0

[5 rows x 14 columns]

2. Drop the columns which are unique for all users like IDs (5 points)

In [0]:
df = dataframe.drop(['CustomerId'], axis=1)
df1 = df.drop(['RowNumber'], axis=1)
df2 = df1.drop(['Surname'], axis=1)

In [90]:
df2.head()

CreditScore Geography  Gender  ...  IsActiveMember  EstimatedSalary  Exited
0          619    France  Female  ...               1        101348.88       1
1          608     Spain  Female  ...               1        112542.58       0
2          502    France  Female  ...               0        113931.57       1
3          699    France  Female  ...               0         93826.63       0
4          850     Spain  Female  ...               1         79084.10       0

[5 rows x 11 columns]

3. Distinguish the feature and target set (5 points)

In [0]:
# split into input (X) and output (Y) variables
X = df2.iloc[:,0:10]
Y = df2.iloc[:,10]


In [92]:
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 10 columns):
CreditScore        10000 non-null int64
Geography          10000 non-null object
Gender             10000 non-null object
Age                10000 non-null int64
Tenure             10000 non-null int64
Balance            10000 non-null float64
NumOfProducts      10000 non-null int64
HasCrCard          10000 non-null int64
IsActiveMember     10000 non-null int64
EstimatedSalary    10000 non-null float64
dtypes: float64(2), int64(6), object(2)
memory usage: 781.3+ KB


In [93]:
X.iloc[:,2]

0       Female
1       Female
2       Female
3       Female
4       Female
5         Male
6         Male
7       Female
8         Male
9         Male
10        Male
11        Male
12      Female
13      Female
14      Female
15        Male
16        Male
17      Female
18        Male
19      Female
20        Male
21      Female
22      Female
23        Male
24      Female
25        Male
26        Male
27        Male
28      Female
29        Male
         ...  
9970      Male
9971    Female
9972      Male
9973      Male
9974      Male
9975      Male
9976    Female
9977    Female
9978      Male
9979    Female
9980      Male
9981      Male
9982    Female
9983      Male
9984      Male
9985      Male
9986      Male
9987      Male
9988      Male
9989      Male
9990      Male
9991    Female
9992      Male
9993      Male
9994    Female
9995      Male
9996      Male
9997    Female
9998      Male
9999    Female
Name: Gender, Length: 10000, dtype: object

In [0]:
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
label_X_gender_encoder = LabelEncoder()
X.iloc[:,2] = label_X_gender_encoder.fit_transform(X.iloc[:,2])

In [0]:

label_X_country_encoder = LabelEncoder()
X.iloc[:,1] = label_X_country_encoder.fit_transform(X.iloc[:,1])

      CreditScore Geography  Gender  ...  HasCrCard  IsActiveMember  EstimatedSalary
0             619    France       0  ...          1               1        101348.88
1             608     Spain       0  ...          0               1        112542.58
2             502    France       0  ...          1               0        113931.57
3             699    France       0  ...          0               0         93826.63
4             850     Spain       0  ...          1               1         79084.10
5             645     Spain       1  ...          1               0        149756.71
6             822    France       1  ...          1               1         10062.80
7             376   Germany       0  ...          1               0        119346.88
8             501    France       1  ...          0               1         74940.50
9             684    France       1  ...          1               1         71725.73
10            528    France       1  ...          0              

In [98]:
countryhotencoder = OneHotEncoder(categorical_features = [1]) # 1 is the country column
X = countryhotencoder.fit_transform(X).toarray()

/usr/local/lib/python3.6/dist-packages/sklearn/preprocessing/_encoders.py:415: FutureWarning: The handling of integer data will change in version 0.22. Currently, the categories are determined based on the range [0, max(values)], while in the future they will be determined based on the unique values.
If you want the future behaviour and silence this warning, you can specify "categories='auto'".
In case you used a LabelEncoder before this OneHotEncoder to convert the categories to integers, then you can now use the OneHotEncoder directly.
  warnings.warn(msg, FutureWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/preprocessing/_encoders.py:451: DeprecationWarning: The 'categorical_features' keyword is deprecated in version 0.20 and will be removed in 0.22. You can use the ColumnTransformer instead.
  "use the ColumnTransformer instead.", DeprecationWarning)


4. Divide the data set into Train and test sets

In [0]:
# Splitting the dataset into the Training and Testing set.
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,Y, test_size = 0.2, random_state = 0)

5. Normalize the train and test data (5 points)

In [0]:
# Feature Scaling
from sklearn.preprocessing import StandardScaler
sc=StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

6. Initialize & build the model (10 points)

In [0]:
from keras.models import Sequential
from keras.layers import Dense

In [0]:
# Initializing the ANN
classifier = Sequential()

In [0]:
# This adds the input layer (by specifying input dimension) AND the first hidden layer (units)
classifier.add(Dense(activation = 'relu', input_dim = 12, units=6, kernel_initializer='uniform'))

In [0]:
# Adding the second hidden layer
# Notice that we do not need to specify input dim. 
classifier.add(Dense(activation = 'relu', units=6, kernel_initializer='uniform')) 

In [0]:
# Adding the output layer
# Notice that we do not need to specify input dim. 
# we have an output of 1 node, which is the the desired dimensions of our output (stay with the bank or not)
# We use the sigmoid because we want probability outcomes
classifier.add(Dense(activation = 'sigmoid', units=1, kernel_initializer='uniform')) 

7. Optimize the model (Optional)

In [0]:
classifier.compile(optimizer='adam', loss = 'binary_crossentropy', metrics=['accuracy'])

In [132]:
classifier.fit(X_train, y_train, batch_size=10, epochs=100)

W0901 16:38:05.779491 139798958815104 deprecation_wrapper.py:119] From /usr/local/lib/python3.6/dist-packages/keras/backend/tensorflow_backend.py:1033: The name tf.assign_add is deprecated. Please use tf.compat.v1.assign_add instead.



Epoch 1/100
8000/8000 [==============================] - 2s 211us/step - loss: 0.4874 - acc: 0.7955
Epoch 2/100
8000/8000 [==============================] - 1s 107us/step - loss: 0.4308 - acc: 0.7960
Epoch 3/100
8000/8000 [==============================] - 1s 107us/step - loss: 0.4263 - acc: 0.7960
Epoch 4/100
8000/8000 [==============================] - 1s 107us/step - loss: 0.4232 - acc: 0.8005
Epoch 5/100
8000/8000 [==============================] - 1s 111us/step - loss: 0.4199 - acc: 0.8211
Epoch 6/100
8000/8000 [==============================] - 1s 110us/step - loss: 0.4171 - acc: 0.8235
Epoch 7/100
8000/8000 [==============================] - 1s 110us/step - loss: 0.4154 - acc: 0.8280
Epoch 8/100
8000/8000 [==============================] - 1s 107us/step - loss: 0.4136 - acc: 0.8309
Epoch 9/100
8000/8000 [==============================] - 1s 109us/step - loss: 0.4121 - acc: 0.8325
Epoch 10/100
8000/8000 [==============================] - 1s 115us/step - loss: 0.4110 - acc: 0.8305

8. Predict the results using 0.5 as a threshold (Optional)

In [134]:
y_pred = classifier.predict_classes(X_test)
print(y_pred)

[[0]
 [0]
 [0]
 ...
 [0]
 [0]
 [0]]


2. Print the Accuracy score and confusion matrix (5 points)

In [135]:
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)
print(cm)

[[1544   51]
 [ 265  140]]


In [141]:
X_test.shape

(2000, 12)

This means that we should have about  (1544+140)=1684  correct classifications out of our total testing data size of  2000 . This means that our accuracy for this trial was  1726÷2000=84.2% , which matches the classifier's prediction

In [136]:
print (((cm[0][0]+cm[1][1])*100)/(cm[0][0]+cm[1][1]+cm[0][1]+cm[1][0]), '% of testing data was classified correctly')

84.2 % of testing data was classified correctly


In [140]:
classifier.evaluate(X_test,y_test)

2000/2000 [==============================] - 0s 46us/step


[0.39819700837135313, 0.842]

Model accuracy 84.2%